### ISSUE TITLE
Static list as `load_default` value results in shared list across all requests

### ISSUE THREAD

**Issue by jlgoolsbee** [marshmallow-code/marshmallow#2114]
Perhaps this isn't a bug so much as a gap in documentation(?), but the result _feels_ like a bug when you're new to Marshmallow, as I am, and when the result seems as inconsistent as shown in the example below. This example is perhaps overly-verbose, but it serves to show how only lists behave in the way described.

```python
#!/usr/bin/env python3.9

import random

from flask import Flask
from flask.views import MethodView
from webargs import fields
from webargs.flaskparser import use_kwargs


app = Flask(__name__)

resource_args = {
    "name": fields.String(load_default="Turboencabulator"),
    "terms": fields.List(fields.String, load_default=["girdlespring"]),
    "parts": fields.List(fields.Dict(keys=fields.String, values=fields.String), load_default=[]),
    "colors": fields.Dict(keys=fields.String, values=fields.String, load_default={}),
    "real": fields.Boolean(load_default=True),
}


def random_color() -> dict:
    return random.choice([
        {"blue": "#00F"},
        {"green": "#0F0"},
        {"red": "#F00"},
    ])


class ResourceView(MethodView):
    decorators = [use_kwargs(resource_args)]

    def post(self, name: str, terms: list[str], parts: list[dict], colors: dict, real: bool):
        name = f"Rockwell {name}"
        terms.extend(["hydrocoptic marzlevanes"])
        parts.extend([{"base-plate": "prefabulated aluminite"}])
        colors = colors | random_color()
        real = False
        return {"name": name, "terms": terms, "parts": parts, "colors": colors, "real": real}
    
    def get(self, name: str, terms: list[str], parts: list[dict], colors: dict, real: bool):
        return {"name": name, "terms": terms, "parts": parts, "colors": colors, "real": real}


app.add_url_rule("/resource", view_func=ResourceView.as_view("resource"))

if __name__ == "__main__":
    app.run(debug=True)
```

In the example, no request body is required for the POST request, all values have a `load_default` defined, and the POST handler (making use of `use_kwargs`) makes modifications to all passed params. After starting this example, an initial POST request with an empty request body will use the `load_default` values for all args; e.g., the response to the first request would be:

```json
{
  "colors": {
    "green": "#0F0"
  },
  "name": "Rockwell Turboencabulator",
  "parts": [
    {
      "base-plate": "prefabulated aluminite"
    }
  ],
  "real": false,
  "terms": [
    "girdlespring",
    "hydrocoptic marzlevanes"
  ]
}
```

... however, subsequent POST requests that result in modifications to lists in the schema permanently modify the contents of that list for that request and all subsequent requests; e.g., the response to the second request would be:

```json
{
  "colors": {
    "blue": "#00F"
  },
  "name": "Rockwell Turboencabulator",
  "parts": [
    {
      "base-plate": "prefabulated aluminite"
    },
    {
      "base-plate": "prefabulated aluminite"
    }
  ],
  "real": false,
  "terms": [
    "girdlespring",
    "hydrocoptic marzlevanes",
    "hydrocoptic marzlevanes"
  ]
}
```

... and a GET request will show the same modification to those lists:

```json
{
  "colors": {},
  "name": "Turboencabulator",
  "parts": [
    {
      "base-plate": "prefabulated aluminite"
    },
    {
      "base-plate": "prefabulated aluminite"
    }
  ],
  "real": true,
  "terms": [
    "girdlespring",
    "hydrocoptic marzlevanes",
    "hydrocoptic marzlevanes"
  ]
}
```

... and subsequent POST requests continue adding elements to the "parts" and "terms" lists, but crucially, as the GET request makes clear, those modifications _only_ affect lists; strings, booleans, and even dictionaries are _not_ persistently modified:

1. The colors dict is empty (after multiple requests one would expect it to have more than one (or all) of the possible results from the `random_color` method)
1. The word "Rockwell" is not repeatedly added to the `name` attribute
3. The `real` boolean isn't permanently flipped to `False`

Searching for related issues is somewhat difficult given the history of the `load_default` parameter being named `missing` in the past, and while I have found a few examples where others use a lambda to init an empty list (e.g. `load_default=lambda: []` or `load_default=lambda: list()`) I've yet to find any documentation or issue that provides that example and/or explains why this behavior only affects lists.

To sum it up: why aren't the other attributes in the schema with `load_default` being persistently modified, and/or is the fact that the lists are persistently modified a bug?

P.S. I've opened this against marshmallow even though the example makes no use of marshmallow directly, as I believe the behavior of the `load_default` and backend `Missing` class stem from webargs use of marshmallow for those features behind the scenes; FWIW, my original example used marshmallow schemas instead of a dict of request arguments, and I saw the same behavior.

**Comment by lafrech** [comment]
This is the "mutable argument" Python gotcha.

Don't pass mutable variables as default arguments to functions.

Marshmallow allows you to pass a callable, so pass a lambda returning 
the static list.

-- 
Jérôme

**Comment by lafrech** [comment]
Duplicate of #1217.

**Comment by jlgoolsbee** [comment]
> Duplicate of #1271.

I'm not seeing the relation to #1271? 😅

I don't necessarily disagree with your comment, but I do still think there's a point of confusion here even if it's not a bug. Perhaps it was just a poor assumption on my part, but from all the examples and docs I'd read, I was under the impression that that request arguments marshaled through the schema and passed to the view – including the defaults supplied by the schema itself – would be tied/scoped to the request, and in practice that _seems_ to be true for everything _except_ lists of strings/objects... 🤷

**Comment by lafrech** [comment]
Oops. I meant #1217.

I think it is not true but you only see it when being caught by a mutable argument.

**Comment by ThiefMaster** [comment]
I wonder if marshmallow shouldn't warn you if you pass any of the common mutable objects (ie list/dict/set) as `load_default`.
Or alternatively, make a copy when using the default, I can't imagine any reason why someone would want to use the same object every time the field/schema is used...

Even as someone who's perfectly aware of the mutable function argument "issue" (or feature, there are some valid usecases for it) in Python I noticed a bunch of `load_default=[]` in my codebase, all in code written or reviewed by me...

`load_default=list` or `load_default=lambda: []` would of course work, but the former is not super obvious on what its purpose is, and the latter takes way more space than just `load_default=[]`...

**Comment by lafrech** [comment]
I believe this is caught by flake8-bugbear as B006.

https://pypi.org/project/flake8-bugbear/

I understand the point and open to discussion but I'm reluctant to add code here if static analysis can catch the problem.

**Comment by ThiefMaster** [comment]
B006 is just for default values in function definitions. Looking at [this failed run](https://github.com/indico/indico/actions/runs/5454780127/jobs/9925432945?pr=5833) from a PR in my repo that adds flake8-bugbear it does not detect the `load_default` values.

**Comment by lafrech** [comment]
My bad. You're right. I got confused.

**Comment by lafrech** [comment]
I'd rather add a warning than copy the input (even if I don't see a legit use case either for a list that would be common to all loads). Not only for perf but also for the 0.0001% who actually rely on this.

(Or does someone want to create a flake8-marshmallow plugin? Just throwing the idea as a joke but that could make sense if a few other rules were identified.)

**Comment by ThiefMaster** [comment]
I'd love to see such a plugin (and eventually support in [ruff](https://github.com/astral-sh/ruff) as well, but those guys seem to be pretty proactive in integrating existing flake8 plugins), especially if there are more pitfalls that could be easily detected...
